source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -124,6 +135,23 @@ Network::IoResult SslSocket::doRead(Buffer::Instance& read_buffer) {
   return {action, bytes_read, end_stream};
 }
 
+void SslSocket::complete(Envoy::Ssl::PrivateKeyOperationStatus status) {
+  ASSERT(async_handshake_in_progress_);
+  async_handshake_in_progress_ = false;
+
+  if (status == Envoy::Ssl::PrivateKeyOperationStatus::Success) {
+    ENVOY_CONN_LOG(debug, "async handshake complete", callbacks_->connection());
+    if (!handshake_complete_) {
```

Comment:
Can this method be ever called with `handshake_complete_ == true`? Perhaps this should be an `ASSERT()` instead?

Reply:
Ok, changed to `ASSERT()`.

Reply:
I mean, changed `async_handshake_in_progress_` check to `ASSERT()`. I think it's possible that (depending on how the private key method provider is implemented) there's a race between the async callback and a `doHandshake()` call from `doWrite()` or `doRead()`. In that case it could be that the handshake is already completed and an ASSERT for `handshake_complete_` would be wrong.